### PR TITLE
[Tidy] Refactor Vizro-AI class

### PR DESCRIPTION
**Pull request by nadijagraca**
## Description
Vizro-ai class refactoring includes:

- setting a default model `ChatOpenAI(model_name="gpt-3.5-turbo-0613", temperature=0)` - TODO: replace with new model - possibly "gpt-3.5-turbo-0125" as the current one will be deprecated on June 13, 2024.
- model can be a string/langchain model and if not specified users will get the above mentioned default model

Possibilities:

1. model not specified 
`vizro_ai = VizroAI()`
2. Langchain model customized by users
```
llm = ChatOpenAI(
    model_name="gpt-3.5-turbo-1106", 
    temperature=1
)
vizro_ai = VizroAI(model=llm)
```
3. Model defined by the user - default temperature in this case will be 0.7 (default temperature of ChatOpenAI)
```
llm = ChatOpenAI(
    model_name="gpt-3.5-turbo-1106" # default temp
)
vizro_ai = VizroAI(model=llm)
```
4. Model name provided as string - default temperature will be 0
```
model = "gpt-3.5-turbo-1106"
vizro_ai = VizroAI(model=model)
```
TO DO:

- [ ] unit tests for the model constructor?
- [x] test all pre-defined models
- [x] change the default
- [ ] check what happens if the user configures a model that is not suited for this type of task ("chat completion"), is the error message clear enough?

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

File: vizro-ai/src/vizro_ai/chains/_llm_models.py [hunk 1]
```diff
@@ -56,33 +35,25 @@ class ModelConstructor:
...
     """
 
-    models: Dict[str, LLM]
-
-    def __init__(self):
-        """Initializes the model manager with a set of predefined LLMs."""
-        self.models = {model["name"]: LLM(**model) for model in PREDEFINED_MODELS}
-
-    def get_llm_model(self, model_name: str, temperature: float = 0) -> LLM_MODELS:
+    def get_llm_model(self, model=None) -> LLM_MODELS:
```
Comment (huong-li-nguyen) [inline vizro-ai/src/vizro_ai/chains/_llm_models.py:38]: Isn't this a breaking change? People who have been using these args before - it wouldn't work for them anymore right? I don't know the context of it, but why do we introduce this breaking change and remove these args?

Comment (huong-li-nguyen) [inline vizro-ai/src/vizro_ai/chains/_llm_models.py:38]: Can we provide typing to `model`? What does it expect?  Still a str? Literal?

File: vizro-ai/src/vizro_ai/_vizro_ai.py [hunk 2]
```diff
@@ -19,41 +18,34 @@ class VizroAI:
     pipeline_manager: PipelineManager = PipelineManager()
     _return_all_text: bool = False
 
-    def __init__(self, model_name: str = "gpt-3.5-turbo-0613", temperature: int = 0):
+    def __init__(self, model=None):
```
Comment (huong-li-nguyen) [inline vizro-ai/src/vizro_ai/_vizro_ai.py:21]: Add type

**Review by huong-li-nguyen — COMMENTED** [review]
I'll let @maxschulz-COL do the actual review as I don't have enough context on this one. But I was just wondering if we consciously introduce the breaking change here? If yes, I think this needs to be highlighted in the changelog at least

File: vizro-ai/src/vizro_ai/_vizro_ai.py [hunk 3]
```diff
@@ -19,41 +18,34 @@ class VizroAI:
     pipeline_manager: PipelineManager = PipelineManager()
```
Comment (maxschulz-COL) [inline vizro-ai/src/vizro_ai/_vizro_ai.py:18]: We should get rid of the `model_constructor` now, as I think it is not needed anymore.

**Review by maxschulz-COL — COMMENTED** [review]
Looking good, this really simplifies things. Added some suggestions.

Some further comments:
- can we add unit tests for the model constructor
- have you tested all pre-defined models?
- we should change the default as you say
- what happens if the user configures a model that is not suited for this type of task ("chat completion"), is the error message clear enough?

File: vizro-ai/src/vizro_ai/chains/_llm_models.py [hunk 4]
```diff
@@ -56,33 +35,25 @@ class ModelConstructor:
...
         """Fetches and initializes an instance of the LLM.
 
-        Args:
-            model_name (str): The name of the LLM.
-            temperature (int, optional): A parameter for the wrapper. Defaults to 0.
-
-        Returns:
+        Returns
```
[the remark this answers is not in the document]
Comment (nadijagraca) [reply vizro-ai/src/vizro_ai/chains/_llm_models.py:41]: For some reason, every time I run `hatch run lint` the colons get removed. I've added the colons, and will not run lint before i push.

**Comment by maxschulz-COL** [comment]
> > * what happens if the user configures a model that is not suited for this type of task ("chat completion"), is the error message clear enough?
> 
> All our predefined models are suited for chat completion, and if user defines model that is not suited for this task, lets say "gpt-3.5-turbo-instruct" they will receive a message "Model gpt-3.5-turbo-instruct not found!".

Yes but I meant what if he configures his own model, not if he chooses one of the pre-defined ones with `string`

**Comment by nadijagraca** [comment]
> > > * what happens if the user configures a model that is not suited for this type of task ("chat completion"), is the error message clear enough?
> > 
> > 
> > All our predefined models are suited for chat completion, and if user defines model that is not suited for this task, lets say "gpt-3.5-turbo-instruct" they will receive a message "Model gpt-3.5-turbo-instruct not found!".
> 
> Yes but I meant what if he configures his own model, not if he chooses one of the pre-defined ones with `string`

They will get: 
`NotFoundError: Error code: 404 - {'error': {'message': 'This is not a chat model and thus not supported in the v1/chat/completions endpoint. Did you mean to use v1/completions?', 'type': 'invalid_request_error', 'param': 'model', 'code': None}}`

**Comment by maxschulz-COL** [comment]
> > > > * what happens if the user configures a model that is not suited for this type of task ("chat completion"), is the error message clear enough?
> > > 
> > > 
> > > All our predefined models are suited for chat completion, and if user defines model that is not suited for this task, lets say "gpt-3.5-turbo-instruct" they will receive a message "Model gpt-3.5-turbo-instruct not found!".
> > 
> > 
> > Yes but I meant what if he configures his own model, not if he chooses one of the pre-defined ones with `string`
> 
> They will get: `NotFoundError: Error code: 404 - {'error': {'message': 'This is not a chat model and thus not supported in the v1/chat/completions endpoint. Did you mean to use v1/completions?', 'type': 'invalid_request_error', 'param': 'model', 'code': None}}`

I am fine with that, what do you think?

File: vizro-ai/changelog.d/20240415_223857_nadija_ratkusic_graca_refactor_vizroai_class.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+<!--
+### Highlights ✨
+
+- A bullet item for the Highlights ✨ category with a link to the relevant PR at the end of your entry, e.g. Enable feature XXX ([#1](https://github.com/mckinsey/vizro/pull/1))
+
+-->
+
+### Removed
+
+- Removed `temperature` argument of `VizroAI` class. ([#423](https://github.com/mckinsey/vizro/pull/423))
```
Comment (maxschulz-COL) [inline vizro-ai/changelog.d/20240415_223857_nadija_ratkusic_graca_refactor_vizroai_class.md:16]: Do we not also remove the `model_name` argument? We should also provide a guide (or link to a guide) with what the user should do now.

File: vizro-ai/src/vizro_ai/chains/_llm_models.py [hunk 6]
```diff
@@ -1,88 +1,59 @@
...
-            return model.wrapper(model_name=model.name, temperature=temperature)
-        else:
-            raise ValueError(f"Model {model_name} not found!")
+    if not model:
+        return ChatOpenAI(model_name=DEFAULT_MODEL, temperature=DEFAULT_TEMPERATURE)
+    if isinstance(model, ChatOpenAI):
+        return model
+    if isinstance(model, str) and model in PREDEFINED_MODELS:
+        return PREDEFINED_MODELS.get(model)["wrapper"](model_name=model, temperature=DEFAULT_TEMPERATURE)
+    raise ValueError(f"Model {model} not found!")
```
[the remark this answers is not in the document]
Comment (maxschulz-COL) [reply vizro-ai/src/vizro_ai/chains/_llm_models.py:55]: Also, can we mention where the user can find the pre-defined models:
- in the docs
- but can the user also find it in code, by importing maybe some private variable?

**Review by maxschulz-COL — COMMENTED** [review]
Looking great, just a few more small things I think

**Review by stichbury — APPROVED** [review]
LGTM 🏆

File: vizro-ai/docs/pages/explanation/faq.md
Comment (maxschulz-COL) [inline vizro-ai/docs/pages/explanation/faq.md:1]: Can we move this into the "How-to" guides, and make a "How to configure Vizro-AI" guide out of it.

Then rewrite it as:
- default usage
- customize with string, plus available strings
- customize entirely with an example

Thanks!

File: vizro-ai/docs/pages/explanation/faq.md [hunk 7]
```diff
@@ -22,23 +24,37 @@ We are working on supporting more models and more vendors. Stay tuned!
...
     Users are also recommended to review the third party API key section of the [disclaimer](../explanation/disclaimer.md) documentation.
 
 ## What parameters does Vizro-AI support?
 Currently, Vizro-AI supports the following parameters:
 
-- `temperature`: A parameter for tuning the randomness of the output. It is set to 0 by
-  default. We recommend setting it to 0 for Vizro-AI usage, as it's mostly
-  deterministic.
-- `model_name`: The name of the model to use. See above for [models currently supported by Vizro-AI](#which-llms-are-supported-by-vizro-ai) for the models supported.
+- `model`: The name of the model or the instantiated model to use. See above for [models currently supported by Vizro-AI](#which-llms-are-supported-by-vizro-ai) for the models supported.
```
Comment (maxschulz-COL) [inline vizro-ai/docs/pages/explanation/faq.md:34]: I have found another instance of `model_name` in one other guide - could you please double check that we have updated the entire documentation?

File: vizro-ai/changelog.d/20240415_223857_nadija_ratkusic_graca_refactor_vizroai_class.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+<!--
+### Highlights ✨
+
+- A bullet item for the Highlights ✨ category with a link to the relevant PR at the end of your entry, e.g. Enable feature XXX ([#1](https://github.com/mckinsey/vizro/pull/1))
+
+-->
+
+### Removed
+
+- Removed `temperature` and `model_name` arguments of `VizroAI` class. For current configuration visit [VizroAI docs](https://vizro.readthedocs.io/projects/vizro-ai/en/latest/pages/explanation/faq/#what-parameters-does-vizro-ai-support) ([#423](https://github.com/mckinsey/vizro/pull/423))
```
Comment (maxschulz-COL) [inline vizro-ai/changelog.d/20240415_223857_nadija_ratkusic_graca_refactor_vizroai_class.md:16]:
```suggestion
- Removed `temperature` and `model_name` arguments of `VizroAI` class. For current configuration options, visit the [Vizro-AI docs](https://vizro.readthedocs.io/projects/vizro-ai/en/latest/pages/explanation/faq/#what-parameters-does-vizro-ai-support) ([#423](https://github.com/mckinsey/vizro/pull/423))
```

Comment (maxschulz-COL) [reply vizro-ai/changelog.d/20240415_223857_nadija_ratkusic_graca_refactor_vizroai_class.md:16]: Make sure this link still holds once we reshuffle the one page as suggested below

File: vizro-ai/changelog.d/20240415_223857_nadija_ratkusic_graca_refactor_vizroai_class.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+### Added
+
+- Enable customization of LLM models provided to VizroAI class. ([#423](https://github.com/mckinsey/vizro/pull/423))
```
Comment (maxschulz-COL) [inline vizro-ai/changelog.d/20240415_223857_nadija_ratkusic_graca_refactor_vizroai_class.md:20]:
```suggestion
- Enable customization of LLM models provided to Vizro-AI class. ([#423](https://github.com/mckinsey/vizro/pull/423))
```

File: vizro-ai/docs/pages/explanation/faq.md
Comment (stichbury) [reply vizro-ai/docs/pages/explanation/faq.md:1]: Can we ticket this separately and do it in a different PR as I'm moving things around (slightly) in Vizro-AI docs as part of #426 and can't face another merge. I'll tackle after those changes!

File: vizro-ai/src/vizro_ai/chains/_llm_models.py [hunk 10]
```diff
@@ -1,88 +1,65 @@
...
-        else:
-            raise ValueError(f"Model {model_name} not found!")
+    if not model:
+        return ChatOpenAI(model_name=DEFAULT_MODEL, temperature=DEFAULT_TEMPERATURE)
+    if isinstance(model, ChatOpenAI):
+        return model
+    if isinstance(model, str) and model in PREDEFINED_MODELS:
+        return PREDEFINED_MODELS.get(model)["wrapper"](model_name=model, temperature=DEFAULT_TEMPERATURE)
+    raise ValueError(
+        f"Model {model} not found! List of available model can be found at https://vizro.readthedocs.io/projects/vizro-ai/en/latest/pages/explanation/faq/#which-llms-are-supported-by-vizro-ai"
```
[the remark this answers is not in the document]
Comment (stichbury) [reply vizro-ai/src/vizro_ai/chains/_llm_models.py:60]: If we change the section, the `--strict` flag will fail the docs build so we'll spot it and fix it just to get that to pass. So probably not necessary?

File: vizro-ai/docs/pages/explanation/faq.md [hunk 11]
```diff
@@ -7,6 +7,8 @@ Vizro-AI currently supports OpenAI models as follows:
 - gpt-4-0613
 - gpt-3.5-turbo-1106 (under testing)
 - gpt-4-1106-preview (under testing, not suitable for production use)
+- gpt-3.5-turbo-0125
```
[the remark this answers is not in the document]
Comment (nadijagraca) [reply vizro-ai/docs/pages/explanation/faq.md:10]: All have been tested. I've added two new ones. We default to `gpt-3.5-turbo` which currently points to `gpt-3.5-turbo-0125`

File: vizro-ai/docs/pages/explanation/faq.md
Comment (nadijagraca) [reply vizro-ai/docs/pages/explanation/faq.md:1]: I've opened a new ticket for this task, and assigned myself to it. [741](https://github.com/McK-Internal/vizro-internal/issues/741)

@stichbury once you merge your PR, I'll start working on adding this ticket.

**Comment by nadijagraca** [comment]
> > > > > * what happens if the user configures a model that is not suited for this type of task ("chat completion"), is the error message clear enough?
> > > > 
> > > > 
> > > > All our predefined models are suited for chat completion, and if user defines model that is not suited for this task, lets say "gpt-3.5-turbo-instruct" they will receive a message "Model gpt-3.5-turbo-instruct not found!".
> > > 
> > > 
> > > Yes but I meant what if he configures his own model, not if he chooses one of the pre-defined ones with `string`
> > 
> > 
> > They will get: `NotFoundError: Error code: 404 - {'error': {'message': 'This is not a chat model and thus not supported in the v1/chat/completions endpoint. Did you mean to use v1/completions?', 'type': 'invalid_request_error', 'param': 'model', 'code': None}}`
> 
> I am fine with that, what do you think?

I'm also fine with the message, it states it is not chat model and not supported.

File: vizro-ai/docs/pages/explanation/faq.md [hunk 12]
```diff
@@ -22,23 +24,38 @@ We are working on supporting more models and more vendors. Stay tuned!
...
         ```py linenums="1"
         from vizro_ai import VizroAI
 
-        vizro_ai = VizroAI(model_name="gpt-3.5-turbo-0613", temperature=0)
+        vizro_ai = VizroAI(model="gpt-3.5-turbo-0613")
+        ```
+
+## How can I customize my model?
+
+`model` parameter of `VizroAI` accepts instantiated model, allowing users to customize it as needed.
```
Comment (maxschulz-COL) [inline vizro-ai/docs/pages/explanation/faq.md:47]:
```suggestion
The `model` parameter of `VizroAI` accepts an instantiated `ChatOpenAI` model, allowing users to customize it as needed.
```

File: vizro-ai/src/vizro_ai/_vizro_ai.py [hunk 13]
```diff
@@ -15,45 +15,37 @@
 class VizroAI:
     """Vizro-AI main class."""
 
-    model_constructor: ModelConstructor = ModelConstructor()
     pipeline_manager: PipelineManager = PipelineManager()
     _return_all_text: bool = False
 
-    def __init__(self, model_name: str = "gpt-3.5-turbo-0613", temperature: int = 0):
+    def __init__(self, model: Union[ChatOpenAI, str] = None):
```
Comment (maxschulz-COL) [inline vizro-ai/src/vizro_ai/_vizro_ai.py:21]:
```suggestion
    def __init__(self, model: Optional[Union[ChatOpenAI, str]] = None):
```

File: vizro-ai/src/vizro_ai/chains/_llm_models.py [hunk 14]
```diff
@@ -1,88 +1,65 @@
...
+DEFAULT_MODEL = "gpt-3.5-turbo"
+DEFAULT_TEMPERATURE = 0
 
-    Attributes
-        name (str): The name of the LLM.
-        max_tokens (int): The maximum number of tokens that the LLM can handle.
-        wrapper (callable): The langchain function used to instantiate the model.
 
-    """
+def _get_llm_model(model: Union[ChatOpenAI, str] = None) -> LLM_MODELS:
```
Comment (maxschulz-COL) [inline vizro-ai/src/vizro_ai/chains/_llm_models.py:40]:
```suggestion
def _get_llm_model(model: Optional[Union[ChatOpenAI, str]] = None) -> LLM_MODELS:
```

**Review by maxschulz-COL — APPROVED** [review]
LGTM now! 💪  Congrats on your first `vizro-ai` PR.

I spotted two more mentions of `model_name` - once in the Jupyer notebook example, and once in the tests..

Can we please still fix those, and could you please double check that there are no more by performing a search of `model_name` through the `vizro-ai` directory.

File: vizro-ai/docs/pages/explanation/faq.md
Comment (stichbury) [reply vizro-ai/docs/pages/explanation/faq.md:1]: @nadijagraca If you want to start on #741 don't let me hold you up. My PR still needs one approval from @maxschulz-COL but it is all ready to go, so you can safely branch from it and work if you want to get started.

https://github.com/mckinsey/vizro/pull/426

**Pull request merged**